crates/core_simd/src/array.rs

```diff
@@ -0,0 +1,244 @@
+use crate::intrinsics;
+use crate::masks::*;
+use crate::vector::ptr::{SimdConstPtr, SimdMutPtr};
+use crate::vector::*;
+
+/// A representation of a vector as an "array" with indices, implementing
+/// operations applicable to any vector type based solely on "having lanes",
+/// and describing relationships between vector and scalar types.
+pub trait SimdArray<const LANES: usize>: crate::LanesAtMost32
```

Comment:
Looking this over again, shouldn't the lane count be an associated const?

```rust
pub trait SimdArray {
    const LANES: usize;
}
```

It's impossible to implement `SimdArray` with different lengths for a single type.

Reply:
Hm... is that a bug?

Reply:
I am OK with using an assoc const fwiw, I just genuinely don't know which one would actually be best here.

Reply:
The associated const would let you do something like `<T as SimdArray>::LANES` which I could see being useful.  No easy way to extract the lane count otherwise, I think.

Reply:
Also it follows the best practice of "if you want to allow this type to implement the trait multiple times, use an associated ~~type~~ const, otherwise use a ~~type~~ const parameter".

Reply:
I can add the assoc const on its own but removing the const from the trait definition induces a cycle in evaluating the well-formedness of the const generic, so I think it can't be done.

Reply:
I'll check it out in a bit, just to see if I can convince it.  I don't see why it shouldn't work.

Reply:
cc @rust-lang/project-const-generics since we seem to be running into an amusing niche case in our expectations of how const things work and interact with Rust programming idioms, vs. how they do in practice.

for note: my attempt was to bound everything currently described by LANES on `Self::LANES` and instead write stuff like
```rust
impl SimdArray for Simd{Ty}<LANES> {
    const LANES: usize = LANES;
}
```
and I experimented with a few different configurations of where-bounds and the like, but I kept hitting cycles or outright failures to type check.

Reply:
After finding that the problem is probably in the bounds not being moved mostly into the implementations of the traits, we will probably catch up to refactoring this in after a bit more tinkering around.